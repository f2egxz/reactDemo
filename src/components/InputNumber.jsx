// -*- mode: react -*-
// -*- coding: utf-8 -*-

/**
 *InputNumber.jsx
 *
 * 数字输入框组件
 *
 * -TODO 数字输入框，可以设置最大值、最小值、每次改变步数、初始值、禁用
 *
 * @author fe.gxz
 */

import React              from 'react';
import { classnames }     from 'util/view-extra'
import Styles             from 'style/InputNumber.css'

/**
 * InputNumber
 *
 * @param {number} min,
 * @param {number} max,
 * @param {number} value,
 * @param {number} step,
 * @param {bool} disabled
 *
 */
class InputNumber extends React.Component {
  constructor(props){
    super()
    this.clickdown = this.clickdown.bind(this)
    this.clickup = this.clickup.bind(this);
    this.changeInput = this.changeInput.bind(this)
    this.state= {
      lastValue:props.value,
      value:props.value,
      show:false,
      focus:false,
      inputNum:props.value,
    }
  }
  componentDidUpdate(){
    if(this.state.lastValue!==this.state.value){
      const {onChange} = this.props;
      onChange(this.state.value)
    }
  }
  render(){
    const {disabled} =this.props;
    return (
      <div className={classnames(Styles.inputNumber,disabled?Styles.disabled:'',this.state.focus?Styles.inputNumbeActive:'')}
           onKeyDown={this.onKeyDown.bind(this)}
           onMouseEnter={this.showarrow.bind(this)}
           onMouseLeave={this.hidearrow.bind(this)}
           onFocus={this.onfocus.bind(this)}
           onBlur={this.onBlur.bind(this)}
      >
        <input type="text" className={classnames(Styles.input,disabled?Styles.disabled:'')}
               value={this.state.inputNum}   onChange={this.changeInput}/>
        {this.state.show&&!disabled?<div className={classnames(Styles.arrowbox)}>
            <span className={classnames(Styles.arrow,Styles.arrowup)}
                  onClick={this.clickup}
            >▲</span>
            <span className={classnames(Styles.arrow ,Styles.arrowdown)}
                  onClick={this.clickdown}
            >▼</span>
          </div>:''}
      </div>
    )
  }
  changeInput(e){
    const num = e.target.value;
    this.setState({
      inputNum:num,
    })
  }

  getPrecision (){
    const {step} = this.props
    const stepString = step+''
    let precision = 0;
    if (stepString.indexOf('.') >= 0) {
      precision = stepString.length - stepString.indexOf('.') - 1;
    }
    return precision
  }

  toPrecisionAsStep(num) {
    if (isNaN(num) || num === '') {
      return num;
    }
    const precision = this.getPrecision();
    return Number(Number(num).toFixed(precision));
  }

  clickup(){
    const {max,step} = this.props;
    const num = this.toPrecisionAsStep(this.state.value+step);
    this.state.value+step<max?
      this.setState({
        inputNum:num,
        value:num,
        lastValue:this.state.value
      }):''
  }

  clickdown(){
    const {min,step} = this.props
    const num = this.toPrecisionAsStep(this.state.value-step)
    this.state.value-step>min?
    this.setState({
      inputNum:num,
      value:num,
      lastValue:this.state.value
    }):''
  }
  onKeyDown(e) {
    if (e.keyCode === 38) {
      this.clickup(e);
    } else if (e.keyCode === 40) {
      this.clickdown(e);
    }
  }
  showarrow(){
    this.setState({
      show:true,
      lastValue:this.state.value
    })
  }
  hidearrow(){
    this.setState({
      show:false,
      lastValue:this.state.value
    })
  }
  onfocus(){
    this.setState({
      focus:true
    })
  }
  onBlur(){
    const num =this.state.inputNum;
    let value;
    isNaN(Number(num))? value = this.state.value : value = Number(num);
    this.setState({
      inputNum:value,
      value:value,
      focus:false
    })
  }
}
InputNumber.defaultProps = {
  min:-Infinity,
  max:Infinity,
  value:'',
  step:1,
  disabled:false
}

InputNumber.PropTypes = {
  min:      React.PropTypes.number,
  max:      React.PropTypes.number,
  value:    React.PropTypes.number,
  step:     React.PropTypes.number,
  disabled: React.PropTypes.bool
}


export default InputNumber

