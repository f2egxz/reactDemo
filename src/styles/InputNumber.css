/*-*--mode: css--*-*/
/*-*--coding: utf-8--*-*/

.input{
  box-sizing: border-box;
  width: 100%;
  text-align: left;
  outline: 0;
  -moz-appearance: textfield;
  line-height: 26px;
  height: 26px;
  transition: all 0.3s linear;
  color: rgba(0, 0, 0, 0.65);
  border: 0;
  border-radius: 4px;
  padding: 0 7px;
}
.inputNumber {
  position: relative;
  margin: 0;
  padding: 0;
  font-size: 12px;
  height: 28px;
  display: inline-block;
  border: 1px solid #d9d9d9;
  border-radius: 4px;
  width: 85px;
  cursor: text;
  color: rgba(0, 0, 0, 0.65);
}
.inputNumbeActive {
  border-color: #49a9ee;
}
.inputNumber:hover {
  border-color: #49a9ee;
}
.arrowbox {
  border-left: 1px solid #d9d9d9;
  width: 22px;
  height: 100%;
  background: #fff;
  position: absolute;
  top: 0;
  right: 0;
  opacity: 1;
  border-radius: 0 4px 4px 0;
  display: flex;
  flex-direction:column;
}

.arrow {
  text-align: center;
  height: 50%;
  overflow: hidden;
  color: rgba(0, 0, 0, 0.43);
  position: relative;
  transition: all 0.1s linear;
  display: block;
  width: 100%;
  font-weight: bold;
  line-height: 100%;
}
.arrowdown {
  border-top: 1px solid #d9d9d9;
  top: -1px;
  cursor: pointer;
}

.arrowup {
  cursor: pointer;
}

.arrowup:hover {
  height: 22px;
  color: #49a9ee;
}

.arrowdown:hover {
  height: 16px;
  margin-top: -2px;
  color: #49a9ee;
}

.disabled {
  background-color: #f7f7f7;
  opacity: 1;
  cursor: not-allowed;
  color: rgba(0, 0, 0, 0.25);
}
.disabled :hover {
  border-color: #e2e2e2;
}
