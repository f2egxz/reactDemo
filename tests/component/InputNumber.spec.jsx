import React from 'react';
import InputNumber from 'component/InputNumber.jsx'
import render from 'react-test-renderer'

test('InputNumber',() => {
  const inputchange = jest.fn();
  const component = render.create(
    <InputNumber value={2} step={1} max={6} min={-3} onChange = {inputchange}/>
  )
  let InputNumberString = component.toJSON();
  expect(InputNumberString).toMatchSnapshot();
})
