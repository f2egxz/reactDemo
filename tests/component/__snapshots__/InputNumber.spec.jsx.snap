exports[`test InputNumber 1`] = `
<div
  className="inputNumber  "
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}>
  <input
    className="input "
    onChange={[Function]}
    type="text"
    value={2} />
  
</div>
`;
